sky/provision/__init__.py

```diff
@@ -0,0 +1,90 @@
+"""This module provides a standard low-level interface that all
+providers supported by Skypilot need to follow."""
+
+import typing
+import functools
+import importlib
+import inspect
+
+from sky.provision import common
+
+
+def _router(func):
+
+    @functools.wraps(func)
+    def _wrapper(*args, **kwargs):
+        # check the signature to fail early
+        inspect.signature(func).bind(*args, **kwargs)
```

Comment:
no, but it should be checked by mypy statically when calling these functions.

I do not use abstract class because this is literally a **singleton API interface**. They work like URLs. Just imagine how you are going to write Flask web apps, where you would not use an abstract class for URL routers. The interface of cloud implementation is actually different (they do not have the `provider_name` field).

originally I also uses an abstract class, then I find it cumbersome to integrate with the centralized service, because you need a general URL-like API (i.e., `provider_name` which is meaningless for individual cloud is a must), instead of an abstract class for all cloud-specific implementation.

Reply:
we had discussed it in a meeting